File: storage/nodejs/dynamo_s3.js
var AWS = require("aws-sdk");
const dotenv = require('dotenv');
const UserDB = require("./objects/user");
const Photo = require("./objects/photo");
dotenv.config();

/* Configurando credenciales */
AWS.config.update({
    accessKeyId: process.env.ACCESS_KEY_ID,
    secretAccessKey: process.env.SECRET_ACCESS_KEY,
    region: process.env.REGION_NAME,
});

/* Clientes de AWS */
var client_dynamodb = new AWS.DynamoDB.DocumentClient();
var client_s3 = new AWS.S3();

/* Metadata */
const table_users = {
    Name: 'Users',
    Attributes: ['Username', 'Password', 'FullName']
}
const table_photos = {
    Name: 'Photos',
    Attributes: ['PhotoURL', 'AlbumName', 'Username']
}
const bucket_name = "practica1.g10.imagenes"

function returnErr(error) {
    console.log(error);
    return false;
}

/* REGISTRAR UN USUARIO */
function add_user(username, password, fullname, base64_photo, filename_photo) {
    // Verificar que no exista el usuario
    return new Promise((resolve, reject) => {
        client_dynamodb.get(
            { TableName: table_users.Name, Key: { 'Username': username } },
            (err, data) => {
                if (err) console.log(err);
                if (data.Item) resolve(returnErr("User already exists"));
                // Creación de parámetros
                const item_users = {
                    Username: username,
                    Password: password,
                    FullName: fullname
                }
                const url = "Fotos_Perfil/" + username + "/actual/" + filename_photo
                const item_photos = {
                    PhotoURL: url,
                    AlbumName: "Fotos de Perfil",
                    Username: username
                }
                console.log("Adding user:", item_users);
                // Insertar usuario
                client_dynamodb.put({
                    TableName: table_users.Name, Item: item_users
                }, function (err) {
                    if (!err) {
                        if (base64_photo && filename_photo) {
                            // Guardar la foto en bucket S3
                            let b64_decode = new Buffer.from(base64_photo, 'base64');
                            const params = {
                                Bucket: bucket_name,
                                Key: url,
                                Body: b64_decode,
                                ContentType: "image"
                            };
                            client_s3.putObject(params, function (err) {
                                if (!err) {
                                    // Insertar ruta en Dynamo
                                    client_dynamodb.put({
                                        TableName: table_photos.Name, Item: item_photos
                                    }, (err) => {
                                        if (!err) {
                                            console.log("User added correctly");
                                            resolve(true);
                                        }
                                    });
                                }
                                else resolve(returnErr(err));
                            });
                        }
                    } else resolve(returnErr(err));
                });
            });
    });
}

/* VERIFICAR LOGIN */
function login_user(username, password) {
    return new Promise((resolve, reject) => {
        const key = { 'Username': username };
        client_dynamodb.get({ TableName: table_users.Name, Key: key },
            function (err, user) {
                if (!err) {
                    if (!user.Item) resolve(returnErr("The user doesn't exists"));
                    // console.log(user)
                    const password_db = user.Item.Password;
                    if (password_db == password) { // Ambas password deben estar en md5
                        console.log("Login is correct");
                        resolve(true);
                    }
                    else resolve(returnErr("The password is incorrect"));
                }
                else resolve(returnErr(err));
            });
    });
}

/* OBTENER TODA LA DATA DE UN USUARIO */
function get_user(__username) {
    return new Promise((resolve, reject) => {
        try {
            const key = { 'Username': __username };
            // Obtener el usuario
            client_dynamodb.get({ TableName: table_users.Name, Key: key },
                function (err, user_db) {
                    if (err) resolve(returnErr(err));
                    const username = user_db.Item.Username;
                    const password = user_db.Item.Password;
                    const fullname = user_db.Item.FullName;
                    let user_response = new UserDB(username, password, fullname);
                    // Obtener sus fotos
                    client_dynamodb.scan({
                        TableName: table_photos.Name, FilterExpression: 'Username=:name', ExpressionAttributeValues: { ":name": key.Username }
                    }, function (err, photos_db) {
                        if (err) resolve(returnErr(err));
                        for (let i = 0; i < photos_db.Items.length; i++) {
                            const photo = photos_db.Items[i];
                            const url = photo.PhotoURL;
                            const albumName = photo.AlbumName;
                            user_response.addPhoto(url, albumName);
                        }
                        console.log(user_response.printUser());
                        resolve(user_response);
                    });
                });
        } catch (error) {
            resolve(returnErr(error));
        }
    });
}

/* SUBIR FOTO A ALBUM DE USUARIO (PARA CREAR UN ALBUM ES NECESARIO SUBIR UNA FOTO) */
function uploadPhoto(username, albumName, base64_photo, filename_photo) {
    return new Promise((resolve, reject) => {
        const url = "Fotos_Publicadas/" + username + "/" + albumName + "/" + filename_photo;
        const item_photos = {
            PhotoURL: url,
            AlbumName: albumName,
            Username: username
        };
        let b64_decode = new Buffer.from(base64_photo, 'base64');
        const params = {
            Bucket: bucket_name,
            Key: url,
            Body: b64_decode,
            ContentType: "image"
        };
        // Insertar ruta en Dynamo
        client_dynamodb.put(
            { TableName: table_photos.Name, Item: item_photos },
            function (err) {
                if (!err) {
                    // Guardar la foto en bucket S3
                    client_s3.putObject(params, function (err) {
                        if (!err) {
                            console.log("Upload Successful")
                            resolve(true);
                        }
                        else resolve(returnErr(err));
                    });
                }
                else resolve(returnErr(err));
            });
    });
}


/* ACTUALIZAR LA FOTO DE PERFIL */
function updateProfilePhoto(__username, new_b64_profile_photo, new_filename_photo) {
    return new Promise((resolve, reject) => {
        // Obtener el usuario
        get_user(__username).then((user) => {
            if (user) {
                let old_photo = user.getProfilePhoto();
                let new_url_old_photo = "Fotos_Perfil/" + __username + "/" + old_photo.getFilename();
                let params = {
                    Bucket: bucket_name,
                    CopySource: bucket_name + '/' + old_photo.getUrl(),
                    Key: new_url_old_photo
                };
                // Copiar a la carpeta común
                client_s3.copyObject(params,
                    function (err) {
                        if (!err) {
                            // Eliminar la que estaba en /actual
                            client_s3.deleteObject({ Bucket: bucket_name, Key: old_photo.getUrl() },
                                function (err) {
                                    if (!err) {
                                        // Cargar la nueva en /actual
                                        let new_url = "Fotos_Perfil/" + __username + "/actual/" + new_filename_photo;
                                        let item_photos = {
                                            PhotoURL: new_url,
                                            AlbumName: "Fotos de Perfil",
                                            Username: __username
                                        };
                                        let b64_decode = new Buffer.from(new_b64_profile_photo, 'base64');
                                        params = {
                                            Bucket: bucket_name,
                                            Key: new_url,
                                            Body: b64_decode,
                                            ContentType: "image"
                                        };
                                        // Guardar la foto en bucket
                                        client_s3.putObject(params, function (err) {
                                            if (!err) {
                                                // Insertar ruta en Dynamo
                                                client_dynamodb.put({ TableName: table_photos.Name, Item: item_photos },
                                                    function (err) {
                                                        if (!err) {
                                                            // Actualizar item antiguo en Dynamo
                                                            let key = {
                                                                PhotoURL: old_photo.getUrl(),
                                                                Username: __username
                                                            };
                                                            client_dynamodb.delete({ TableName: table_photos.Name, Key: key },
                                                                function (err) {
                                                                    if (!err) {
                                                                        // Reutilizar el item cambiando la url
                                                                        item_photos.PhotoURL = new_url_old_photo;
                                                                        client_dynamodb.put({ TableName: table_photos.Name, Item: item_photos },
                                                                            function (err) {
                                                                                if (!err) {
                                                                                    console.log("Profile photo updated correctly");
                                                                                    resolve(true);
                                                                                } else resolve(returnErr(err));
                                                                            });
                                                                    } else resolve(returnErr(err));
                                                                });
                                                        } else resolve(returnErr(err));
                                                    });
                                            } else resolve(returnErr(err));
                                        });
                                    } else resolve(returnErr(err));
                                });
                        } else { console.log("F", params); resolve(returnErr(err)); }
                    });
            } else resolve(returnErr("User not found"));
        });
    });
}

/* EDITAR UN USUARIO (FULLNAME O USERNAME) */
function updateUser(__username, __password, new_username, new_fullname) {
    return new Promise((resolve, reject) => {
        const key = { 'Username': __username };
        // Obtener el usuario
        client_dynamodb.get({ TableName: table_users.Name, Key: key },
            function (err, user_db) {
                if (err) resolve(returnErr(err));
                let password_db = user_db.Item.Password;
                let fullname_db = user_db.Item.FullName;

                if (password_db != __password) {  // Ambas password deben estar en md5
                    resolve(returnErr("The password is incorrect"));
                }
                if (new_username) {  // Se deben actualizar todas URLS
                    // Obtener el usuario
                    get_user(__username).then((user) => {
                        updateUsername_URLS(user, new_username).then((r) => {
                            if (r) {
                                client_dynamodb.delete({ TableName: table_users.Name, Key: key },
                                    function (err) {
                                        if (!err) {
                                            add_user((new_username || __username), __password, (new_fullname || fullname_db), '', '')
                                                .then((r) => resolve(r));
                                        } else resolve(returnErr(err));
                                    });
                            } else resolve(returnErr("Error: URLS didn't updated"))
                        });
                    });
                } else {
                    // Sólo se cambia el nombre
                    client_dynamodb.delete({ TableName: table_users.Name, Key: key },
                        function (err) {
                            if (!err) {
                                add_user((new_username || __username), __password, (new_fullname || fullname_db), '', '')
                                    .then((r) => resolve(r));
                            } else resolve(returnErr(err));
                        });
                }
            });
    });
}

/* ACTUALIZAR TODAS LAS RUTAS POR EL CAMBIO DE USERNAME */
function updateUsername_URLS(old_user, new_username) {
    return new Promise((resolve, reject) => {
        let response = false;
        // Actualizar en el Bucket S3
        let old_photos = old_user.getPhotos();
        for (let i = 0; i < old_photos.length; i++) {
            const old_photo = old_photos[i];
            let new_photo = new Photo(old_photo.getUrl(), old_photo.getAlbumName(), old_photo.getUsername());
            new_photo.changeUsername(new_username);
            const copy_params = {
                Bucket: bucket_name,
                CopySource: bucket_name + '/' + old_photo.getUrl(),
                Key: new_photo.getUrl()
            };
            // Copiar a la nueva carpeta
            client_s3.copyObject(copy_params, function (err) {
                if (!err) {
                    // Eliminar la que estaba en /actual
                    client_s3.deleteObject({ Bucket: bucket_name, Key: old_photo.getUrl() },
                        function (err) {
                            if (!err) {
                                // Insertar ruta en Dynamo
                                const item_photos = {
                                    PhotoURL: new_photo.getUrl(),
                                    AlbumName: new_photo.getAlbumName(),
                                    Username: new_photo.getUsername()
                                };
                                client_dynamodb.put({ TableName: table_photos.Name, Item: item_photos },
                                    function (err) {
                                        if (!err) {
                                            // Eliminar ruta antigua en Dynamo
                                            const key = {
                                                PhotoURL: old_photo.getUrl(),
                                                Username: old_photo.getUsername()
                                            }
                                            client_dynamodb.delete({ TableName: table_photos.Name, Key: key },
                                                function (err) {
                                                    if (!err) {
                                                        response = true;
                                                    } else resolve(returnErr(err));
                                                });
                                        } else resolve(returnErr(err));
                                    });
                            } else resolve(returnErr(err));
                        });
                } else resolve(returnErr(err));
            });
        }
        if (response)
            console.log("URLS updated")
        resolve(response);
    });
}


/* ELIMINAR UN ALBUM (No se debe poder eliminar el album de fotos de perfil) */
function deleteAlbum(username, albumName) {
    return new Promise((resolve, reject) => {
        let response = false;
        get_user(username).then((user) => {
            let photos = user.getPhotos();
            for (let i = 0; i < photos.length; i++) {
                const photo = photos[i];
                if (photo.getAlbumName() == albumName) {
                    // Eliminar en bucket S3
                    client_s3.deleteObject({ Bucket: bucket_name, Key: photo.getUrl() },
                        function (err) {
                            if (!err) {
                                // Eliminar en DynamoDB
                                const key = {
                                    PhotoURL: photo.getUrl(),
                                    Username: photo.getUsername()
                                };
                                client_dynamodb.delete({ TableName: table_photos.Name, Key: key },
                                    function (err) {
                                        if (!err) {
                                            response = true;
                                        }
                                        else resolve(returnErr(err));
                                    });
                            }
                            else resolve(returnErr(err));
                        });
                }
            }
            if (response)
                console.log("Album deleted");
            resolve(response);
        });
    });
}

/* ELIMINAR UNA FOTO DEL ALBUM */
function deletePhoto(username, URL_photo) {
    return new Promise((resolve, reject) => {
        // Eliminar en bucket S3 (Sin la dirección 's3.amazonaws.com/practica1-G10-imagenes/')
        client_s3.deleteObject({ Bucket: bucket_name, Key: URL_photo },
            function (err) {
                if (!err) {
                    // Eliminar en DynamoDB
                    const key = {
                        PhotoURL: URL_photo,
                        Username: username
                    };
                    client_dynamodb.delete({ TableName: table_photos.Name, Key: key },
                        function (err) {
                            if (!err) {
                                console.log("Photo deleted");
                                resolve(true);
                            }
                            else resolve(returnErr(err));
                        }
                    );
                }
                else resolve(returnErr(err));
            }
        );
    });
}

// add_user('luisd', '0000', 'Luis Danniel Castellanos', getBase64('../perfil1.txt'), 'img1.jpg');
// login_user('luisd', '0000');
// uploadPhoto('luisd', 'Pensums', getBase64('../sistemas.txt'), 'Sistemas.jpg');
// uploadPhoto('luisd', 'Pensums', getBase64('../industrial.txt'), 'Industrial.jpg');
// updateProfilePhoto('luisd', getBase64('../perfil2.txt'), 'img2.jpg');
// deletePhoto('luisd','Fotos_Publicadas/luisd/Pensums/Industrial.jpg');
// deleteAlbum('ldecast', 'Pensums');
// updateUser('luisd', '0000', '', 'Luis Danniel Ernesto Castellanos Galindo');
// updateUser('luisd', '0000', 'ldecast', '');

// get_user('ldecast');












function getBase64(path) {
    var fs = require('fs');
    const data = fs.readFileSync(path, 'utf8');
    return data;
}

File: storage/python/dynamo_s3.py
import base64
from objects.photo import Photo
from objects.user import UserDB
import boto3  # pip install boto3
from dotenv import load_dotenv
import os
import io

load_dotenv()

ACCESS_KEY_ID = os.getenv('ACCESS_KEY_ID')
SECRET_ACCESS_KEY = os.getenv('SECRET_ACCESS_KEY')
REGION_NAME = os.getenv('REGION_NAME')

client_dynamodb = None
client_s3 = None
client_rekognition = None

table_users = {
    'Name': 'Users',
    'Attributes': ['Username', 'Password', 'FullName']
}

table_photos = {
    'Name': 'Photos',
    'Attributes': ['PhotoURL', 'Tags', 'Username']
}

bucket_name = "practica1.g10.imagenes"


def connect_AWS_Services() -> bool:
    try:
        # CONECTAR A LA BASE DE DATOS
        global client_dynamodb
        client_dynamodb = boto3.client('dynamodb', aws_access_key_id=ACCESS_KEY_ID,
                                       aws_secret_access_key=SECRET_ACCESS_KEY, region_name=REGION_NAME)
        # CONECTAR A BUCKET S3 DE IMAGENES
        global client_s3
        client_s3 = boto3.client('s3',
                                 aws_access_key_id=ACCESS_KEY_ID,
                                 aws_secret_access_key=SECRET_ACCESS_KEY,
                                 region_name=REGION_NAME)
        # CONECTAR A REKOGNITION
        global client_rekognition
        client_rekognition = boto3.client('rekognition', aws_access_key_id=ACCESS_KEY_ID,
                                          aws_secret_access_key=SECRET_ACCESS_KEY, region_name=REGION_NAME)
    except:
        print("Something went wrong connecting with AWS Services")
        return False
    else:
        print("AWS Services running!")
        return True


# REGISTRAR UN USUARIO
def add_user(username: str, password: str, fullname: str, base64_photo: str, filename_photo: str) -> bool:
    # Verificar que no exista el usuario
    if 'Item' in client_dynamodb.get_item(TableName=table_users['Name'], Key={'Username': {'S': username}}):
        print("User already exists")
        return False

    item_users = {
        'Username': {'S': username},
        'Password': {'S': password},
        'FullName': {'S': fullname}
    }
    url = "Fotos_Perfil/"+username+"/actual/"+filename_photo
    item_photos = {
        'PhotoURL': {'S': url},
        # TODO Agregar extracción de etiquetas de la foto de perfil
        'Tags': {'SS': ["Person", "Human"]},
        'Username': {'S': username}
    }
    try:
        print("Adding user:", item_users)
        # Insertar usuario
        client_dynamodb.put_item(
            TableName=table_users['Name'], Item=item_users)
        if base64_photo and filename_photo:
            # Guardar la foto en bucket S3
            b64_decode = base64.b64decode(base64_photo)
            client_s3.upload_fileobj(io.BytesIO(b64_decode), bucket_name, url,
                                     ExtraArgs={'ContentType': "image"})
            # Insertar ruta en Dynamo
            client_dynamodb.put_item(
                TableName=table_photos['Name'], Item=item_photos)
    except:
        print("The user has not been added")
        return False
    else:
        print("User added correctly")
        return True


# VERIFICAR LOGIN
def login_user(username: str, password: str, photo: str) -> bool:
    key = {
        'Username': {'S': username}
    }
    user = client_dynamodb.get_item(TableName=table_users['Name'], Key=key)
    if 'Item' not in user:
        print("The user doesn't exists")
        return False
    if password:
        password_db = user['Item']['Password']['S']
        if password_db == password:  # Ambas password deben estar en md5
            print("Login is correct")
            return True
        print("The password is incorrect")
        return False
    elif photo:
        # Insertar reconocimiento facial
        pass
    else:
        print("Password or photo must be provided.")
        return False


# OBTENER TODA LA DATA DE UN USUARIO
def get_user(__username: str) -> UserDB:
    key = {
        'Username': {'S': __username}
    }
    # Obtener el usuario
    user_db = client_dynamodb.get_item(TableName=table_users['Name'], Key=key)
    username = user_db['Item']['Username']['S']
    password = user_db['Item']['Password']['S']
    fullname = user_db['Item']['FullName']['S']
    user_response = UserDB(username, password, fullname)
    # Obtener sus fotos
    photos_db = client_dynamodb.scan(
        TableName=table_photos['Name'], FilterExpression='Username=:name', ExpressionAttributeValues={":name": key['Username']})
    for photo in photos_db['Items']:
        url = photo['PhotoURL']['S']
        tags = photo['Tags']['SS']
        user_response.addPhoto(url, tags)
    # print(user_response)
    return user_response


# OBTENER LAS ETIQUETAS DE UNA FOTO
def getPhotoLabels(b64_decode: bytes):
    image = {'Bytes': b64_decode}
    maxLabels = 5
    response = client_rekognition.detect_labels(
        Image=image, MaxLabels=maxLabels
    )
    labels = []
    for label in response['Labels']:
        labels.append(label['Name'])
    # print(labels)
    return labels


# SUBIR FOTO Y ALMACENAR SUS ETIQUETAS
def uploadPhoto(username: str, base64_photo: str, filename_photo: str):
    url = "Fotos_Publicadas/"+username+"/"+filename_photo
    b64_decode = base64.b64decode(base64_photo)
    item_photos = {
        'PhotoURL': {'S': url},
        'Tags': {'SS': getPhotoLabels(b64_decode)},
        'Username': {'S': username}
    }
    # Insertar ruta en Dynamo
    client_dynamodb.put_item(
        TableName=table_photos['Name'], Item=item_photos)
    # Guardar la foto en bucket
    client_s3.upload_fileobj(io.BytesIO(b64_decode), bucket_name, url,
                             ExtraArgs={'ContentType': "image"})
    print("Upload Successful")


# ACTUALIZAR LA FOTO DE PERFIL
def updateProfilePhoto(__username: str, new_b64_profile_photo: str, new_filename_photo: str) -> bool:
    # Obtener el usuario
    user = get_user(__username)
    old_photo = user.getProfilePhoto()
    new_url_old_photo = "Fotos_Perfil/"+__username+"/"+old_photo.getFilename()
    copy_source = {
        'Bucket': bucket_name,
        'Key': old_photo.getUrl()
    }
    # Copiar a la carpeta común
    client_s3.copy(
        copy_source,
        bucket_name,
        new_url_old_photo
    )
    # Eliminar la que estaba en /actual
    client_s3.delete_object(
        Bucket=bucket_name,
        Key=old_photo.getUrl()
    )
    # Cargar la nueva en /actual
    new_url = "Fotos_Perfil/"+__username+"/actual/"+new_filename_photo
    b64_decode = base64.b64decode(new_b64_profile_photo)
    item_photos = {
        'PhotoURL': {'S': new_url},
        'Tags': {'SS': getPhotoLabels(b64_decode)},
        'Username': {'S': __username}
    }
    # Guardar la foto en bucket
    client_s3.upload_fileobj(io.BytesIO(b64_decode), bucket_name, new_url,
                             ExtraArgs={'ContentType': "image"})
    # Insertar ruta en Dynamo
    client_dynamodb.put_item(
        TableName=table_photos['Name'], Item=item_photos)
    # Actualizar item antiguo en Dynamo
    key = {
        'PhotoURL': {'S': old_photo.getUrl()},
        'Username': {'S': __username}
    }
    client_dynamodb.delete_item(
        TableName=table_photos['Name'], Key=key)
    # Reutilizar el item cambiando la url
    item_photos['PhotoURL']['S'] = new_url_old_photo
    client_dynamodb.put_item(
        TableName=table_photos['Name'], Item=item_photos)
    return True


# EDITAR UN USUARIO (FULLNAME O USERNAME)
def updateUser(__username: str, __password: str, new_username: str, new_fullname: str) -> bool:
    key = {'Username': {'S': __username}}
    # Obtener el usuario
    user_db = client_dynamodb.get_item(TableName=table_users['Name'], Key=key)
    password_db = user_db['Item']['Password']['S']
    fullname_db = user_db['Item']['FullName']['S']

    if password_db != __password:  # Ambas password deben estar en md5
        print("The password is incorrect")
        return False

    if new_username:  # Se deben actualizar todas URLS
        # Obtener el usuario
        user = get_user(__username)
        updateUsername_URLS(user, new_username)

    client_dynamodb.delete_item(
        TableName=table_users['Name'],
        Key=key
    )
    add_user((new_username or __username), __password,
             (new_fullname or fullname_db), '', '')
    return True


# ACTUALIZAR TODAS LAS RUTAS POR EL CAMBIO DE USERNAME
def updateUsername_URLS(old_user: UserDB, new_username: str):
    # Actualizar en el Bucket S3
    for old_photo in old_user.getPhotos():
        new_photo = Photo(
            old_photo.getUrl(), old_photo.getTags, old_photo.getUsername()
        )
        new_photo.changeUsername(new_username)
        copy_source = {
            'Bucket': bucket_name,
            'Key': old_photo.getUrl()
        }
        # Copiar a la nueva carpeta
        client_s3.copy(
            copy_source,
            bucket_name,
            new_photo.getUrl()
        )
        # Eliminar la que estaba en /actual
        client_s3.delete_object(
            Bucket=bucket_name,
            Key=old_photo.getUrl()
        )
        # Insertar ruta en Dynamo
        item_photos = {
            'PhotoURL': {'S': new_photo.getUrl()},
            'Tags': {'SS': new_photo.getTags()},
            'Username': {'S': new_photo.getUsername()}
        }
        client_dynamodb.put_item(
            TableName=table_photos['Name'], Item=item_photos)
        # Eliminar ruta antigua en Dynamo
        key = {
            'PhotoURL': {'S': old_photo.getUrl()},
            'Username': {'S': old_photo.getUsername()}
        }
        client_dynamodb.delete_item(
            TableName=table_photos['Name'],
            Key=key
        )


# ELIMINAR UNA FOTO
def deletePhoto(username: str, URL_photo: str) -> bool:
    # Eliminar en bucket S3
    client_s3.delete_object(
        Bucket=bucket_name,
        Key=URL_photo  # Sin la dirección 's3.amazonaws.com/practica1-G10-imagenes/'
    )
    # Eliminar en DynamoDB
    key = {
        'PhotoURL': {'S': URL_photo},
        'Username': {'S': username}
    }
    client_dynamodb.delete_item(
        TableName=table_photos['Name'],
        Key=key
    )
    return True


# Con fines de testeo
def getBase64(path: str) -> str:
    text_file = open(path, "r")
    data = text_file.read()
    text_file.close()
    return data


if __name__ == '__main__':
    if connect_AWS_Services():
        # add_user("luisd", "0000", "Luis Danniel Castellanos",
        #          getBase64('../testing/perfil1.txt'), "img1.jpg")
        # updateUser("luisd", "0000", "ldecast", "Luis Danniel Ernesto Castellanos Galindo")
        uploadPhoto('luisd', getBase64(
            '../testing/sistemas.txt'), 'Sistemas.jpg')
        # updateProfilePhoto("ldecast", getBase64('../testing/perfil2.txt'), 'img2.jpg')
        # add_user("luisd", "1234", "LuisDa", getBase64('../testing/perfil1.txt'), "hola.jpg")
        # deleteAlbum('ldecast','Pensums')
        # deletePhoto('ldecast','Fotos_Perfil/ldecast/img1.jpg')
        # get_user('luisd')
        pass
